source/distributed/tria.cc

```diff
@@ -3491,32 +3491,42 @@ namespace parallel
+      // cells that are not owned locally. However, in the worst case we want
+      // to connect to a vertex that is 'dim' hops away from the locally owned
+      // cell. Depending on the order of the periodic face map, we might
+      // connect to that point by chance or miss it. However, after looping
+      // through the faces once per space dimension we can be sure that all
+      // connections to vertices have been made.
+      for (unsigned int i=0; i<dim; ++i)
+        for (it = this->get_periodic_face_map().begin(); it!= this->get_periodic_face_map().end(); ++it)
+          {
+            const cell_iterator &cell_1 = it->first.first;
+            const unsigned int face_no_1 = it->first.second;
+            const cell_iterator &cell_2 = it->second.first.first;
+            const unsigned int face_no_2 = it->second.first.second;
+            const std::bitset<3> &face_orientation = it->second.second;
 
-          if (cell_1->level() == level &&
-              cell_2->level() == level)
-            {
-              for (unsigned int v=0; v<GeometryInfo<dim-1>::vertices_per_cell; ++v)
-                {
-                  // take possible non-standard orientation of faces into account
-                  const unsigned int vface0 =
-                    GeometryInfo<dim>::standard_to_real_face_vertex(v,face_orientation[0],
-                                                                    face_orientation[1],
-                                                                    face_orientation[2]);
-                  if (marked_vertices[cell_1->face(face_no_1)->vertex_index(vface0)] ||
-                      marked_vertices[cell_2->face(face_no_2)->vertex_index(v)])
-                    marked_vertices[cell_1->face(face_no_1)->vertex_index(vface0)]
-                      = marked_vertices[cell_2->face(face_no_2)->vertex_index(v)]
-                        = true;
-                }
-            }
-        }
+            if (cell_1->level() == level &&
+                cell_2->level() == level)
+              {
+                for (unsigned int v=0; v<GeometryInfo<dim-1>::vertices_per_cell; ++v)
+                  {
+                    // take possible non-standard orientation of faces into account
+                    const unsigned int vface0 =
+                      GeometryInfo<dim>::standard_to_real_face_vertex(v,face_orientation[0],
+                                                                      face_orientation[1],
+                                                                      face_orientation[2]);
+                    if (marked_vertices[cell_1->face(face_no_1)->vertex_index(vface0)] ||
+                        marked_vertices[cell_2->face(face_no_2)->vertex_index(v)])
+                      marked_vertices[cell_1->face(face_no_1)->vertex_index(vface0)]
+                        = marked_vertices[cell_2->face(face_no_2)->vertex_index(v)]
+                          = true;
```

Comment:
@bangerth Regarding the repetition factor, have a look here: we construct the relevant vertex marking by telling both vertices along the current face pair that they should be marked active if any of the two has a locally owned cell adjacent to it. Or if it was marked by a face pair that came before it. Before this patch, we had to be lucky (and we indeed were for smaller cases) that the lower ids would get marked because some other part of the grid was locally active.

Also, it is not the number of cells that matters here, but the number of periodicity constraints that create such a connection. And there are at most `dim` hops in terms of periodicity, as many as there are periodic directions in the mesh.

As a matter of fact, it would have been enough to do `(dim+1)/2` loop iterations because the `periodic_face_map` variable contains the pair `(cell_1, cell_2)` also the other way around `(cell_2,cell_1)`, which ensures that we cover all hops over two dimensions and thus this case could not happen in 2D. However, there can be 3 hops in 3D. I still decided to stay with `dim` loops to be more stable with respect to the implementation of the `periodic_face_map` and be conservative in general. This part of the code takes so little time we can afford looping twice.